Fix memory leak in PS and BS.

shitz opened the pull request:
A flaw in how segment verification was handled when receiving path segments through Zookeeper lead to a memory leak.

On startup of a server it can happen that the first path segments come from Zookeeper. If not all certificates are available the server requests it from the local CS with the cache only flag set. However, the CS doesn't have this certificate yet, which lead to a rerequesting loop. This loop also prevented requesting the certs from the appropriate beacon servers when new path segments were arriving.

Fixed the issue by having newer path segments always overwrite older ones in the unverified segments map.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/netsec-ethz/scion/1159)
<!-- Reviewable:end -->

ercanucan commented:
@shitz what's the status on this PR? is it going to be merged soon so that we can deploy?

kormat commented:
Reviewed 1 of 3 files at r1, 1 of 1 files at r2.
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[python/scion_elem/scion_elem.py, line 193 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1159#-KpL2Xk1hvuTHmZoIsiR:-KpL2Xk1hvuTHmZoIsiS:b-ykp28a) ([raw file](https://github.com/netsec-ethz/scion/blob/c6b4df5461c0a8c5745e22305f7716b54f4c7e86/python/scion_elem/scion_elem.py#L193)):*
> ```Python
>         else:
>             self._DefaultMeta = UDPMetadata
>         self.unverified_segs = {}
> ```

It probably makes sense to make this an expiring dict, say with the TTL of 1h. Any sgement that you get that can't be verified within an hour is highly unlikely to be useful. This will also prevent any segments that are never verifiable from staying around permanently.

---

*[python/scion_elem/scion_elem.py, line 422 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1159#-KpL3ljcvXveOVGEgUXF:-KpL3ljcvXveOVGEgUXG:b-dxgo2g) ([raw file](https://github.com/netsec-ethz/scion/blob/c6b4df5461c0a8c5745e22305f7716b54f4c7e86/python/scion_elem/scion_elem.py#L422)):*
> ```Python
>             with self.req_trcs_lock:
>                 _, meta = self.requested_trcs.get((isd, ver), (None, None))
>                 # Prefer requesting from remote BS.
> ```

"remote" and "BS" assumptions are incorrect. I'm also not sure what you're trying to do, here.

---

*[python/scion_elem/scion_elem.py, line 431 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1159#-KpL4MRbQhM61P76Jyua:-KpL4MRbQhM61P76Jyub:b-nbf72f) ([raw file](https://github.com/netsec-ethz/scion/blob/70dfac84c0ef5ab1cddbdf5c027f7dc77390fa51/python/scion_elem/scion_elem.py#L431)):*
> seg_meta.meta == self._get_cs()

That's.. a bizarre check. What's the intention?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1159)*
<!-- Sent from Reviewable.io -->

shitz commented:
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[python/scion_elem/scion_elem.py, line 193 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1159#-KpL2Xk1hvuTHmZoIsiR:-KpLQ8k4wpofyx1SADG8:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/c6b4df5461c0a8c5745e22305f7716b54f4c7e86/python/scion_elem/scion_elem.py#L193)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

It probably makes sense to make this an expiring dict, say with the TTL of 1h. Any sgement that you get that can't be verified within an hour is highly unlikely to be useful. This will also prevent any segments that are never verifiable from staying around permanently.
</blockquote></details>

Done.

---

*[python/scion_elem/scion_elem.py, line 422 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1159#-KpL3ljcvXveOVGEgUXF:-KpLT-vvSUOpFT5CSPW5:b-8jw2g5) ([raw file](https://github.com/netsec-ethz/scion/blob/c6b4df5461c0a8c5745e22305f7716b54f4c7e86/python/scion_elem/scion_elem.py#L422)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

"remote" and "BS" assumptions are incorrect. I'm also not sure what you're trying to do, here.
</blockquote></details>

Where is it requested from then?

---

*[python/scion_elem/scion_elem.py, line 431 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1159#-KpL4MRbQhM61P76Jyua:-KpLSaRy07VSw_QRF4LP:b87pmr9) ([raw file](https://github.com/netsec-ethz/scion/blob/70dfac84c0ef5ab1cddbdf5c027f7dc77390fa51/python/scion_elem/scion_elem.py#L431)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

> seg_meta.meta == self._get_cs()

That's.. a bizarre check. What's the intention?
</blockquote></details>

`requested_trcs` contains the outstanding trc requests. The problem is that if the path segment that triggered the TRC request was from zookeeper then `meta` will point to the local CS. Thus you want to issue a new request `seg_meta.meta` does not point to the local CS.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1159)*
<!-- Sent from Reviewable.io -->

kormat commented:
Reviewed 1 of 1 files at r3.
Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[python/scion_elem/scion_elem.py, line 422 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1159#-KpL3ljcvXveOVGEgUXF:-KpOyOa2T3jIoaLP6i9h:b-ixmnsw) ([raw file](https://github.com/netsec-ethz/scion/blob/c6b4df5461c0a8c5745e22305f7716b54f4c7e86/python/scion_elem/scion_elem.py#L422)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

Where is it requested from then?
</blockquote></details>

E.g.:
- path server receives up-segment from a local BS.
- path server receives any segment from remote PS during lookup.

---

*[python/scion_elem/scion_elem.py, line 431 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1159#-KpL4MRbQhM61P76Jyua:-KpOysWcjH5T4r5MPhve:b-h3c1st) ([raw file](https://github.com/netsec-ethz/scion/blob/70dfac84c0ef5ab1cddbdf5c027f7dc77390fa51/python/scion_elem/scion_elem.py#L431)):*
> The problem is that if the path segment that triggered the TRC request was from zookeeper then meta will point to the local CS.

That's a bug. If there is no original meta in the seg_meta, then one should not be added for the local CS. Also you're assuming there's only a single local CS instance, which is bad.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1159)*
<!-- Sent from Reviewable.io -->

shitz commented:
Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[python/scion_elem/scion_elem.py, line 422 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1159#-KpL3ljcvXveOVGEgUXF:-KpPfdzM7VIT-onqgCIe:b-fi6yci) ([raw file](https://github.com/netsec-ethz/scion/blob/c6b4df5461c0a8c5745e22305f7716b54f4c7e86/python/scion_elem/scion_elem.py#L422)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

E.g.:
- path server receives up-segment from a local BS.
- path server receives any segment from remote PS during lookup.
</blockquote></details>

Changed comment entirely.

---

*[python/scion_elem/scion_elem.py, line 431 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1159#-KpL4MRbQhM61P76Jyua:-KpPfgxCp_TZmGCOQcuy:b-u2e77x) ([raw file](https://github.com/netsec-ethz/scion/blob/70dfac84c0ef5ab1cddbdf5c027f7dc77390fa51/python/scion_elem/scion_elem.py#L431)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

> The problem is that if the path segment that triggered the TRC request was from zookeeper then meta will point to the local CS.

That's a bug. If there is no original meta in the seg_meta, then one should not be added for the local CS. Also you're assuming there's only a single local CS instance, which is bad.
</blockquote></details>

Changed according to offline discussion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1159)*
<!-- Sent from Reviewable.io -->

kormat commented:
Reviewed 1 of 2 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[python/scion_elem/scion_elem.py, line 431 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1159#-KpL4MRbQhM61P76Jyua:-KpPtyPvlWqf6HH1pNoM:bxj7r35) ([raw file](https://github.com/netsec-ethz/scion/blob/70dfac84c0ef5ab1cddbdf5c027f7dc77390fa51/python/scion_elem/scion_elem.py#L431)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

Changed according to offline discussion.
</blockquote></details>

This breaks in `_check_trc_reqs`, as it assumes that `requested_trcs` always contains a useful `meta`.

---

*[python/scion_elem/scion_elem.py, line 431 at r4](https://reviewable.io:443/reviews/netsec-ethz/scion/1159#-KpPu6TRaImQNSgxi7KZ:-KpPu6TRaImQNSgxi7K_:b-jqrji3) ([raw file](https://github.com/netsec-ethz/scion/blob/5339f76ee374694759c30b05fa135c59936eb3a0/python/scion_elem/scion_elem.py#L431)):*
> ```Python
>             with self.req_trcs_lock:
>                 _, meta = self.requested_trcs.get((isd, ver), (None, None))
>                 # There is already an outstanding request for the missing TRC.
> ```

This comment should be inside the `if` clause, so it's scope is unambiguous.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1159)*
<!-- Sent from Reviewable.io -->

shitz commented:
Review status: 1 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[python/scion_elem/scion_elem.py, line 431 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1159#-KpL4MRbQhM61P76Jyua:-KpPuywEtkP6teFXhfkP:boj1qgt) ([raw file](https://github.com/netsec-ethz/scion/blob/70dfac84c0ef5ab1cddbdf5c027f7dc77390fa51/python/scion_elem/scion_elem.py#L431)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This breaks in `_check_trc_reqs`, as it assumes that `requested_trcs` always contains a useful `meta`.
</blockquote></details>

Good catch. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1159)*
<!-- Sent from Reviewable.io -->

kormat commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1159)*
<!-- Sent from Reviewable.io -->

shitz commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[python/scion_elem/scion_elem.py, line 431 at r4](https://reviewable.io:443/reviews/netsec-ethz/scion/1159#-KpPu6TRaImQNSgxi7KZ:-KpPvTU_DARFPPUVEpsk:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/5339f76ee374694759c30b05fa135c59936eb3a0/python/scion_elem/scion_elem.py#L431)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This comment should be inside the `if` clause, so it's scope is unambiguous.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1159)*
<!-- Sent from Reviewable.io -->

kormat commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1159#-:-KpPverj5T3EACIicE6x:bnfp4nl)*
<!-- Sent from Reviewable.io -->

kormat commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1159#-:-KpQFxVtC_6a814jmAgR:bnfp4nl)*
<!-- Sent from Reviewable.io -->
